File: C++/fullBinaryTreeToSumBinaryTree.cpp
#include <bits/stdc++.h>

using namespace std;

int sumTree(vector<int> dlr, int l1, int r1, vector<int> &ldr, int l2, int r2)
{
    int temp;
    if (l2 == r2)
    {
        temp = ldr[l2];
        ldr[l2] = 0;
        return temp;
    }

    int root = l1;
    while (ldr[root] != dlr[l1])
        root++;

    int leftSum = sumTree(dlr, l1 + 1, root, ldr, l2, root - 1);
    int rightSum = sumTree(dlr, root + 1, r1, ldr, root + 1, r2);
    temp = ldr[root];
    ldr[root] = leftSum + rightSum;
    return temp + ldr[root];
}

void sumTree2(vector<int> &inorder, int left, int right)
{
    int mid = (left + right) / 2;
    if (mid == left)
    {
        inorder[mid] = 0;
        return;
    }
    inorder[mid] = accumulate(inorder.begin() + left, inorder.begin() + right, -inorder[mid]);
    sumTree2(inorder, left, mid);
    sumTree2(inorder, mid + 1, right);
}

int main(int argc, char const *argv[])
{
    string line;
    getline(cin, line);
    istringstream dlr_stream(line);
    vector<int> dlr((istream_iterator<int>(dlr_stream)), istream_iterator<int>());
    getline(cin, line);
    istringstream ldr_stream(line);
    vector<int> ldr((istream_iterator<int>(ldr_stream)), istream_iterator<int>());

    sumTree(dlr, 0, dlr.size() - 1, ldr, 0, ldr.size() - 1);
    sumTree2(ldr, 0, ldr.size());

    copy(ldr.begin(), ldr.end(), ostream_iterator<int>(cout, " "));
    cout << endl;
    return 0;
}

File: LeetCode/LeetCode-Offer/fullBinaryTreeToSumBinaryTree.php
<?php

function sumBinaryTree($dlr, $l1, $r1, &$ldr, $l2, $r2)
{
    if ($l2 == $r2) {
        $temp = (int) $ldr[$l2];
        $ldr[$l2] = 0;
        return $temp;
    }

    $root = $l1;
    while ($ldr[$root] != $dlr[$l1])
        $root++;

    $leftSum = sumBinaryTree($dlr, $l1 + 1, $root, $ldr, $l2, $root - 1);
    $rightSum = sumBinaryTree($dlr, $root + 1, $r1, $ldr, $root + 1, $r2);
    $temp = (int) $ldr[$root];
    $ldr[$root] = $leftSum + $rightSum;
    return $temp + (int) $ldr[$root];
}

function sumBinaryTree2(array &$ldr, int $left, int $right)
{
    $mid = ($left + $right) >> 1;
    if ($mid == $left) {
        $ldr[$mid] = 0;
        return;
    }
    $temp = -$ldr[$mid];
    for ($i = $left; $i < $right; $i++)
        $temp += $ldr[$i];
    $ldr[$mid] = $temp;
    sumBinaryTree2($ldr, $left, $mid);
    sumBinaryTree2($ldr, $mid + 1, $right);
}

/*
10 -2 8 -4 6 7 5
8 -2 -4 10 7 6 5
 */
$dlr = explode(' ', trim(fgets(STDIN)));
$ldr = explode(' ', trim(fgets(STDIN)));
// $dlr = [10, -2, 8, -4, 6, 7, 5];
// $ldr = [8, -2, -4, 10, 7, 6, 5];
// sumBinaryTree($dlr, 0, count($dlr) - 1, $ldr, 0, count($ldr) - 1);
sumBinaryTree2($ldr, 0, count($ldr));
echo implode(' ', $ldr);
